FDT Memory Detection + Boot Allocator Framework

kjhayes opened the pull request:
Adds a "boot memory allocator" framework, and implements a simple SLAB based memory allocator for use before we can set up a buddy allocator for regular page sized allocations.

Adds new features to the FDT library making reading common properties and getting inherited property info much easier.

Detects the bounds of memory using the FDT and protects reserved regions of memory like the kernel, firmware, and FDT from boot memory allocations.

kjhayes committed:
FDT reserved memory entries

Added basic support to "kernel/of/fdt.h" for reading the memory reservations in the FDT.
These are just blocks of physical memory which we will need to reserve in the allocators because something
special is there we cannot overwrite (we don't actually need to know or care what specifically it is)

kjhayes committed:
Merge branch 'early_output' of github.com:riscy-development/riscy-os into fdt_mem

kjhayes committed:
Basic Boot Memory Allocator

Adds a simple linked "free-list" memory allocator.
Hasn't been tested very closely because the code to give
free regions from the FDT to the memory map hasn't been written yet
(So it can't be tested)

kjhayes committed:
Added More FDT Functionality

Added support for finding nodes by device_type and unit_name (unit_name needs work though)
Also added function to get the parents of a Node, which will be needed to get information
like #address-cells and #interrupt-cells, etc. which is inherited from parents.

kjhayes committed:
FDT Memory Detection

Adds a system for a memory "detection" system to hand off memory to the "boot allocator".
This is done by having the detector call boot_free on all of the non-overlapping regions in physical memory
which are availible for allocations. Then that memory should be able to be allocated by using "boot_alloc"

So new boot-allocator designs could be added as long as they implement boot_alloc and boot_free, without knowing
anything about how memory is actually found (ACPI, multiboot, FDT, etc.)

kjhayes committed:
FDT Inherited Property Support

Adds support functions for getting access to inherited FDT properties
such as "#address-cells" and "#size-cells".

kjhayes committed:
FDT reg Property Functions

Added functions for correctly reading the "reg" property in FDT nodes.
These take into account the affect "#address-cells" and "#size-cells" have on "regs".

kjhayes committed:
Added FDT file extensions to .gitignore

kjhayes committed:
Protect Kernel, Firmware, and FDT from Boot Allocator

Added correctness checks to the FDT memory detection to make sure that
the kernel, firmware and FDT itself are not freed when adding memory to the boot
allocator. With this, barring any unexpected bugs, the boot allocator should be up and running.

kjhayes committed:
Formatted for Pull Request

kjhayes committed:
Merge branch 'early_output' of github.com:riscy-development/riscy-os into fdt_mem

kjhayes committed:
Fixed Spelling Mistakes

kjhayes committed:
Fixing spelling mistakes misformatted?

I hate how pedantic the formatting is Nino

kjhayes committed:
Merge branch 'early_output' of github.com:riscy-development/riscy-os into fdt_mem

kjhayes committed:
Making sure FDT structs don't have padding

kjhayes committed:
Formatting and Spellchecking

egelja commented on .gitignore:
```diff
@@ -504,4 +504,8 @@ $RECYCLE.BIN/
 
+# Device Tree Files
+*.dtb
+*.dts
```
no put this at the top before the `# Start of ...`

egelja commented on include/kernel/mem/boot.h:
```diff
@@ -0,0 +1,62 @@
+#ifndef __KERNEL_MEM_BOOT_H__
+#define __KERNEL_MEM_BOOT_H__
```
```suggestion
#ifndef __KERNEL__MEM_BOOT_H__
#define __KERNEL__MEM_BOOT_H__
```

`__NAMESPACE__FILE_H__`

egelja commented on include/kernel/mem/boot.h:
```diff
@@ -0,0 +1,62 @@
+#ifndef __KERNEL_MEM_BOOT_H__
+#define __KERNEL_MEM_BOOT_H__
+
+#include <kernel/error.h>
```
```suggestion
#include "kernel/error.h"
```

egelja commented on include/kernel/of/fdt.h:
```diff
@@ -8,6 +8,10 @@
 extern "C" {
 #endif
 
+// Bad debugging system (should integrate into Kconfig or delete entirely later)
```
```suggestion
// TODO: Bad debugging system (should integrate into Kconfig or delete entirely later)
```

egelja commented on include/kernel/of/fdt.h:
```diff
@@ -21,7 +25,7 @@ struct fdt {
     uint32_t size_dt_strings;
     uint32_t size_dt_struct;
-};
+} __attribute__((packed));
```
I say we make a header for this

```c
#ifdef __GNUC__
#  define KERNEL_PACKED __attribute__((packed))
#else
#  warning Packed structs are not supported.
#  define KERNEL_PACKED
#endif
```

kjhayes replied:
I'm fine with this but that `#warning` needs to be an `#error`

egelja replied:
sure I wasn't sure so i went with `#warning`

egelja commented on include/kernel/of/fdt_mem.h:
```diff
@@ -0,0 +1,10 @@
+
+#include <kernel/error.h>
+#include <kernel/mem/boot.h>
+#include <kernel/of/fdt.h>
```
these should be quoted includes

egelja commented on include/kernel/of/fdt.h:
can you add in a link in a comment to the schema of an FDT?

and also comment the structs accordingly

kjhayes replied:
The link is a good idea, I'm not commenting structs which are defined according to the spec, it will lead to people trying to use them without reading the spec at all, which is bad, because either we rewrite the entire spec in comments or someone violates it eventually.

egelja replied:
:+1:

egelja commented on kernel/mem/boot/free_list.c:
```diff
@@ -0,0 +1,282 @@
+ * Get the next region (next higher address) in the free list (or NULL)
+ */
+struct boot_free_region* boot_free_list_next(struct boot_free_region* region);
+//
```
shouldn't these be static

egelja commented on kernel/mem/boot/free_list.c:
```diff
@@ -0,0 +1,282 @@
+// How big will the region be if we allocate from the end?
+static ssize_t
+region_size_alloc_after(
+    struct boot_free_region* region, size_t size, unsigned int alignment
```
can we use fixed width types pls

kjhayes replied:
No, it doesn't need to be a fixed width, and 2^65535 is too large of a region to handle by orders of magnitude.

egelja replied:
No I meant make `alignment` a `uint8_t` to make it clear it should be a small value.

kjhayes replied:
I'll make this one `uword_t` when that commit gets merged.

egelja replied:
sure

egelja commented on kernel/mem/boot/free_list.c:
```diff
@@ -0,0 +1,282 @@
+    //        we should split the region in the free list to waste less memory.
+    //        This is especially important for page allocations)
+    size_t unalignment = (end - size) & ((1ull << alignment) - 1ull);
+    region->size -= (size + unalignment);
```
```suggestion
    region->size = region_size_alloc_after(region, size, alignment);
```

kjhayes replied:
Actually no, because the TODO means we shouldn't

kjhayes replied:
We need the alignment specifically to do it

kjhayes replied:
I'll add a helper function for that

egelja commented on kernel/mem/boot/free_list.c:
```diff
@@ -0,0 +1,282 @@
+    uintptr_t start = (uintptr_t)region;
+    // Get how far off from being aligned it is
+    size_t unalignment = (1ull << alignment) - (start & ((1ull << alignment) - 1ull));
+    size_t new_size = region->size - (size + unalignment);
```
```suggestion
    size_t new_size = region_size_alloc_before(region, size, alignment);
```

egelja commented on kernel/mem/boot/free_list.c:
```diff
@@ -0,0 +1,282 @@
+        }
+    }
+    else {
+        // We need to move the region
```
```suggestion
        // We need to move the region start pointer
```

kjhayes replied:
No, the "start pointer" is the region itself, if you move the start pointer you need to update all of the fields in the struct and modify the previous and next in the list. So you're genuinely moving the entire region.

egelja replied:
icic ok makes sense thanks

egelja replied:
ya because the list is ordered

egelja replied:
you should make a note of that invariant next to the free list definition

egelja commented on kernel/mem/boot/free_list.c:
```diff
@@ -0,0 +1,282 @@
+            if (smallest_fit > (size_t)greater) {
+                smallest_fit = (size_t)greater;
+                smallest_fitting = curr;
+                if (before == greater) {
+                    should_alloc_before = true;
+                }
+                else {
+                    should_alloc_before = false;
+                }
```
```suggestion
                should_alloc_before = (before == greater);
```

kjhayes replied:
Actually on review I'm going to swap this, (all things being equal we probably want to do an "after" allocation by default because it's simpler)

kjhayes replied:
The change in setting the bool is fine though

egelja replied:
:+1:

egelja commented on kernel/mem/boot/free_list.c:
```diff
@@ -0,0 +1,282 @@
+
+    struct boot_free_region* curr = boot_free_list_begin();
+    while (curr != NULL) {
+        ssize_t after = region_size_alloc_after(curr, size, alignment);
```
```suggestion
        /* Find which allocation type leaves the region larger */
        ssize_t after = region_size_alloc_after(curr, size, alignment);
```

egelja commented on kernel/mem/boot/free_list.c:
```diff
@@ -0,0 +1,282 @@
+        if (curr_end == start) {
+            // Special case: We can fully merge with the one before
+            struct boot_free_region* next = boot_free_list_next(curr);
+            if (next != NULL && (void*)next == end) {
```
can you make a note somewhere that the free list is assumed to be in-order

egelja commented on kernel/mem/boot/free_list.c:
```diff
@@ -0,0 +1,282 @@
+    // fully_before and/or fully_after should now be populated
+    if (fully_before == NULL && fully_after == NULL) {
+        // This shouldn't happen but let's be safe
+        return KERR_EXIST;
```
```suggestion
        kprintf("This should not occur: %s", __FILE__);
        return KERR_EXIST;
```

egelja commented on kernel/mem/boot/free_list.c:
```diff
@@ -0,0 +1,282 @@
+struct boot_free_region*
+boot_free_list_next(struct boot_free_region* region)
+{
+    return (struct boot_free_region*)region->next;
```
this cast is unnecessary

egelja replied:
`region->next` is already defined as type `struct boot_free_region*`

kjhayes replied:
Eh okay, I don't really care about redundant casts but this one I'm fine removing

egelja commented on kernel/of/fdt.c:
```diff
@@ -13,6 +13,21 @@ fdt_struct_ptr_invalid(struct fdt* fdt, void* ptr)
+    if (ptr < start || ptr >= end) {
+        return true;
+    }
+    return false;
```
```suggestion
    return (ptr < start) || (ptr >= end);
```

egelja commented on kernel/of/fdt.c:
```diff
@@ -285,7 +411,13 @@ fdt_string_from_offset(struct fdt* fdt, size_t offset)
+#ifndef FDT_DEBUG_DEBUG_NAMELESS_NODES
     return node->unit_name;
+#else
+    // If the first character is the terminator, replace it with a debug name
```
well actually only the root node is "nameless"

but since all nodes technically have a prefix of `/`

the root node's name is `/`

egelja replied:
Actually ya

in `kernel.c`

do `printk("/");` before each node

kjhayes replied:
Their names don't actually start with a `/` that's just the way the spec refers to some of them. The root node isn't `/` it is actually nameless the spec uses '/` for the sake of brevity not because that's a true prefix. (Also the debugging wasn't meant for the root node, it's to catch any weird behavior in other nodes)

egelja replied:
Ok ic

egelja commented on kernel/of/fdt.c:
```diff
@@ -115,6 +130,67 @@ fdt_size(struct fdt* fdt)
+fdt_reserve_entry_address(struct fdt_reserve_entry* entry)
+{
+    // TODO: 32-bit systems should use be32toh
+    return (void*)(uintptr_t)be64toh(entry->address);
```
```suggestion
#if CONFIG_64_BIT 
    return (void*)be64toh(entry->address);
#else
    return (void*)be32toh(entry->address);
#endif
```

egelja commented on kernel/of/fdt.c:
```diff
@@ -115,6 +130,67 @@ fdt_size(struct fdt* fdt)
+size_t
+fdt_reserve_entry_size(struct fdt_reserve_entry* entry)
+{
+    // TODO: 32-bit systems should use be32toh
+    return (size_t)be64toh(entry->size);
```
```suggestion
#if CONFIG_64_BIT
    return (size_t)be64toh(entry->size);
#else
    return (size_t)be32toh(entry->size);
#endif
```

kjhayes replied:
Yeah that's fine, I wasn't sure if that macro was removed because it's not used `stdint.h` anymore

kjhayes replied:
Also *GASP* do I see violations of the divine DRY principles??? That's 5 whole lines of code which is the same between two functions... shame on you.

egelja replied:
I'm not trying to go overboard just trying to figure out when copy pasting is ok and when we shouldn't do it

egelja commented on kernel/of/fdt.c:
```diff
@@ -264,6 +348,48 @@ fdt_node_next_subnode(struct fdt* fdt, struct fdt_node* subnode)
+        return 0;
+    }
+
+    struct fdt_node* curr_branch[max_depth + 1]; // C99 feature
```
```suggestion
    struct fdt_node* curr_branch[max_depth + 1];
```

we are using C23

kjhayes replied:
hmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmmm, C23 makes this an optional feature...

kjhayes replied:
I hate that, let me think about what to do to handle this

egelja replied:
We're only using GCC and we pretty much only support 
GCC. 

So it's fine

kjhayes replied:
Yeah but I don't like that this is optional and we're just acting like it's not

kjhayes replied:
It's practically not an issue but bugs me a bit

egelja replied:
Actually C23 makes it mandatory 

> The C23 standard makes VLA types mandatory again.  

https://en.wikipedia.org/wiki/Variable-length_array

kjhayes replied:
This is the description of it I found "Variably-modified types (but not [VLAs](https://en.wikipedia.org/wiki/Variable-length_array) which are [automatic variables](https://en.wikipedia.org/wiki/Automatic_variable) allocated on the stack) become a mandatory feature." (https://en.wikipedia.org/wiki/C23_(C_standard_revision)#cite_note-N2778-25) which would mean this situation is not mandatory.

kjhayes replied:
I can just add the check for the `__STD_C_NO_VLA__` macro so if this ever actually becomes an issue an error will be raised

egelja commented on kernel/of/fdt.c:
```diff
@@ -148,12 +224,20 @@ fdt_node_begin(struct fdt* fdt)
 
 struct fdt_node*
-fdt_next_node(struct fdt* fdt, struct fdt_node* node)
+fdt_next_node(struct fdt* fdt, struct fdt_node* node, int* depth)
```
```suggestion
fdt_next_node(struct fdt* fdt, struct fdt_node* node, size_t* depth)
```

kjhayes replied:
I mention this issue in a different comment

egelja commented on kernel/of/fdt.c:
```diff
@@ -264,6 +348,48 @@ fdt_node_next_subnode(struct fdt* fdt, struct fdt_node* subnode)
+
+    struct fdt_node* curr_branch[max_depth + 1]; // C99 feature
+
+    int depth = 0;
```
```suggestion
    size_t depth = 0;
```

kjhayes replied:
I'll make `depth` a `word_t` when that branch gets merged in. It really should be signed though, because if the tree isn't formed correctly we'll want a negative depth we can check for easily, not an unsigned underflow

egelja replied:
I see. What about a ssize_t then?

kjhayes replied:
It just needs to handle the maximum depth of the device tree, it could easily be handled by a `uint8_t` but since we don't need a fixed size this is an example where `word_t` makes sense

kjhayes replied:
The real `libFDT` uses an `int` for this purpose in their basically equivalent function for context

egelja replied:
Sure then

egelja commented on kernel/of/fdt.c:
```diff
@@ -264,6 +348,48 @@ fdt_node_next_subnode(struct fdt* fdt, struct fdt_node* subnode)
+    // Copy over as much of the branch as we can into the return array
+    // (flipping the order so parents[0] = the direct ancestor of "node"
+    for (size_t i = 0; i < num_to_return; i++) {
+        parents[i] = curr_branch[depth - (int)(i + 1)];
```
```suggestion
        parents[i] = curr_branch[depth - i - 1];
```

kjhayes replied:
Some stupid part of my brain hates chained subtractions like that, I really prefer doing subtraction of an addition. `x - y - z` makes me worried about order of operations and my response to any order of operations issue is just more parenthesis (which I think is the best answer)

egelja replied:
Ok makes sense. Don't need the cast to int tho.

kjhayes replied:
`i` is a `size_t`, depth doesn't need to be so it'll be a `word_t` when we have those in this branch so it'll become `(word_t)(i + 1)`

kjhayes replied:
actually `i` doesn't need to be a `size_t` now that I think about it. Both will be `word_t` when we get them

egelja replied:
Sure

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+size_t
+fdt_prop_num_cells(struct fdt_prop* prop)
+{
+    return fdt_prop_val_len(prop) >> 2;
```
```suggestion
    return fdt_prop_val_len(prop) / sizeof(uint32_t);
```

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+    if (num_cells == 0) {
+        // This should never happen (Corrupted FDT)
+        // TODO: log warning
+        return (uint32_t)0;
```
```suggestion
        return 0;
```

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+    struct fdt_prop* reg = fdt_get_prop_by_name(fdt, node, NULL, "reg");
+
+    if (reg == NULL) {
+        return (size_t)0;
```
```suggestion
        return 0;
```

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+    }
+
+    // 4 bytes per cell
+    size_t reg_cell_len = fdt_prop_val_len(reg) >> 2;
```
```suggestion
    size_t reg_cell_len = fdt_prop_num_cells(reg);
```

kjhayes replied:
Thanks for catching this

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+    }
+
+    uint32_t address_cells = fdt_node_get_address_cells(fdt, node);
+    uint32_t size_cells = fdt_node_get_address_cells(fdt, node);
```
these are the same value

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+
+    switch (address_cells) {
+        case 1:
+            *address = (void*)(uintptr_t)be32toh(*address_ptr);
```
```suggestion
            *address = (void*)be32toh(*address_ptr);
```

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+    return;
+err_exit:
+    *address = NULL;
+    *size = (size_t)0;
```
```suggestion
    *size = 0;
```

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+}
+
+uint32_t
+fdt_node_get_size_cells(struct fdt* fdt, struct fdt_node* node)
```
this is the same code as `fdt_node_get_address_cells` with one different string

make a helper function

kjhayes replied:
The helper function is "fdt_node_get_inherited_prop", both of these give an example of how to use that function. A helper function for this would have to be named `fdt_node_get_first_cell_of_inherited_prop_which_must_exist_and_should_have_a_single_cell` which is not a function I will ever write and I don't think you should either

egelja replied:
`fdt_node_get_first_cell_of_prop`?

kjhayes replied:
That's just `fdt_node_get_cell(node,0)`

kjhayes replied:
The rest of the code is just making sure that property has at least 1 cell

egelja replied:
The verification code could be extracted into a helper

kjhayes replied:
I mean yeah, but at a certain point where the extra helper functions hurt code simplicity, don't sacrifice KISS for DRY and vice versa.

egelja replied:
Ya but there's like 15 LOC for validation

kjhayes replied:
It's just "get array", "get array length", "check access is in bounds", "access element". It's an insanely common pattern but it doesn't mean it should be abstracted away. Unless you want every single array access in the kernel to go through a helper function.

egelja replied:
Just leave it

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+        goto err_exit;
+    }
+
+    uint32_t address_cells = fdt_node_get_address_cells(fdt, node);
+    uint32_t size_cells = fdt_node_get_address_cells(fdt, node);
+    uint32_t reg_block_cells = address_cells + size_cells;
```
this code is copy-pasted (and incorrect in both places)

make a function that returns `reg_block_cells`

kjhayes replied:
I'm patching the two places, but that function would be 3 lines long and only used in these two places. I think DRY is a valuable concept but not for something this simple.

kjhayes replied:
It's literally the sum of two function calls

egelja replied:
then its fine it just was wrong in both places that's why I noticed it

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+    // Size cells come after address cells
+    uint32_t* size_ptr = address_ptr + address_cells;
+
+    uint32_t msig_cell; // Most significant cell in 64-bit
+    uint32_t lsig_cell; // Least significant cell in 64-bit
```
```suggestion
    uint64_t msig_cell; // Most significant cell in 64-bit
    uint64_t lsig_cell; // Least significant cell in 64-bit
```

kjhayes replied:
Each one of these is a single cell (`uint32_t`), which needs to be rejoined together to form the larger number, they're not `uint64_t`'s on their own.

egelja replied:
I know, but look at the next suggestion.

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+            *address = (void*)(uintptr_t)be32toh(*address_ptr);
+            break;
+        case 2:
+            msig_cell = be32toh(address_ptr[0]);
+            lsig_cell = be32toh(address_ptr[1]);
+            *address = (void*)(((uint64_t)msig_cell << 32) | (uint64_t)lsig_cell);
```
```suggestion
            msig_cell = be32toh(address_ptr[0]) << 32;
            lsig_cell = be32toh(address_ptr[1]);
            *address = (void*)(msig_cell | lsig_cell);
```

egelja replied:
I'm sorry but I just see this and think "too many parentheses it must be lisp"

I'm just trying to reduce parentheses

kjhayes replied:
They should be `uint32_t`'s though. And then combined to form one bigger `uint64_t`.

egelja replied:
Well combining two 32 bit ints into a 64 bit one is a common pattern, right?

Why not make a utils file for combining numbers. Like 4x 16 into a 64. Should be all inline.

kjhayes replied:
It's not that common, (this is literally just a `be64toh` written out) admittedly thinking about it I only wrote it like this because I was thinking about the restrictions of PCI ECAM space not the device tree. I'll swap it to just a `be64toh`

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +491,248 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+            break;
+        default:
+            goto err_exit;
+    }
+
+    switch (size_cells) {
+        case 1:
+            *size = (size_t)be32toh(*size_ptr);
+            break;
+        case 2:
+            msig_cell = be32toh(size_ptr[0]);
+            lsig_cell = be32toh(size_ptr[1]);
+            *size = (size_t)(((uint64_t)msig_cell << 32) | (uint64_t)lsig_cell);
+            break;
+        default:
+            goto err_exit;
+    }
```
this code is also copy-pasted

kjhayes replied:
It's also two related but distinct values being computed and should only be done in this single location. DRY is generally good, but blindly adhering to it when it hurts code simplicity/readability it's not worth it.

egelja replied:
I'm well aware I just wanted to make sure this would be the only case

egelja commented on kernel/of/fdt_mem.c:
```diff
@@ -0,0 +1,192 @@
+        }
+    }
+
+    // Should be unreachable
```
```suggestion
    // Should be unreachable
    unreachable();
```

egelja commented on kernel/of/fdt_mem.c:
```diff
@@ -0,0 +1,192 @@
+{
+    // Should be unnecessary (but it's worth checking in case someone messes up)
+    if (size == 0) {
+        return (size_t)0;
```
```suggestion
        return 0;
```

egelja commented on kernel/of/fdt_mem.c:
```diff
@@ -0,0 +1,192 @@
+
+    // Check if this is a region before the end of the kernel (protects the kernel and
+    // firmware which is before it)
+    extern int __kernel_end[];
```
```suggestion
    extern void* __kernel_end;
```

this isn't defined in our linker script though

egelja replied:
also i think you have to take the address of it, like I did with `__init_array_start` in `arch/riscv/kernel.c`

kjhayes replied:
It is defined in the linker script, and it's defined as an array to avoid taking the address. It's of type `int` to avoid people  who don't know how linker script variables work from thinking the "value" of the `void*` is the end of the kernel instead of the symbol itself

kjhayes replied:
This way we also are correct even if people mistakenly try to take `&__kernel_end` because `__kernel_end == &__kernel_end` for array types.

egelja replied:
Alright sure but shouldn't it be declared `uintptr_t` then?

egelja commented on kernel/of/fdt_mem.c:
```diff
@@ -0,0 +1,192 @@
+    if (start < (void*)__kernel_end) {
+        if ((void*)(start + size) <= (void*)__kernel_end) {
+            // Entirely before the kernel
+            return (size_t)0;
```
```suggestion
            return 0;
```

egelja commented on kernel/of/fdt_mem.c:
```diff
@@ -0,0 +1,192 @@
+    }
+    else {
+        // No addresses lower than the end of the kernel (so no overlap)
+    }
```
```suggestion
```

egelja commented on kernel/of/fdt_mem.c:
```diff
@@ -0,0 +1,192 @@
+        else {
+            if (size == 0) {
+                // We got fully trimmed
+                return (size_t)0;
```
```suggestion
                return 0;
```

egelja commented on kernel/of/fdt_mem.c:
```diff
@@ -0,0 +1,192 @@
+            return try_boot_free_region(fdt, start, size)
+                   + try_boot_free_region(fdt, split_start, split_size);
+        }
+        else {
```
you don't need the `else` since you return in the if

egelja commented on kernel/of/fdt_mem.c:
```diff
@@ -0,0 +1,192 @@
+
+        if (trim_or_split_around_reserved(
+                &start, &size, &split_start, &split_size, resv_start, resv_size
+            )) {
```
```suggestion
            ) //
        ) {
```

kjhayes replied:
your formatter did this not me, that being said, I like the first one better (the pointless `//` looks bad)

egelja replied:
i like it on the next line better, and I can live with the `//`

kjhayes replied:
@echavemann settle this

egelja replied:
Wait I have an idea

```c
    bool changed = trim_or_split_around_reserved(
        &start, &size, &split_start, &split_size, resv_start, resv_size
    );

    if (changed) {/* ... */}         
```

kjhayes replied:
Eh, I don't hate it but I'm not going to consistently do it

egelja replied:
i just dont like the indented `{`

egelja replied:
it feels like you're hiding it

egelja commented on kernel/of/fdt_mem.c:
```diff
@@ -0,0 +1,192 @@
+            }
+        }
+
+        entry = fdt_next_reserve_entry(fdt, entry);
```
```suggestion
        }
        // We're still one contiguous region (continue)
        entry = fdt_next_reserve_entry(fdt, entry);
```

egelja commented on kernel/of/fdt_mem.c:
```diff
@@ -0,0 +1,192 @@
+        return try_boot_free_region(fdt, start, size)
+               + try_boot_free_region(fdt, split_start, split_size);
+    }
+    else {
+        if (size == 0) {
+            // We got fully trimmed
+            return (size_t)0;
+        }
+        else {
+            // We're still one contiguous region (continue)
+        }
+    }
```
```suggestion
    if (size == 0) {
        // We got fully trimmed
        return 0;
    }
    
    // We're still one contiguous region (continue)
```

try to avoid nesting whenever possible, it significantly improves readability

kjhayes replied:
I'm good with this I'll make the changes, I was checking exhaustive cases and forgot to go back and remove the elses

egelja commented on kernel/of/fdt_mem.c:
```diff
@@ -0,0 +1,192 @@
+    // We made it through (possibly shrunk)
+    if (boot_free(start, size) != KERR_SUCCESS) {
+        // We couldn't free for some reason
+        return (size_t)0;
```
```suggestion
        return 0;
```

kjhayes replied:
I'm not changing this (or any of the other similar situations), it doesn't hurt anything and is a good reminder to check exact what your return type is and why you're returning the `0`.

egelja replied:
Then why is it done in some places but not others?

kjhayes replied:
Because we can't always be perfect

I said it's `good` not needed. Doing it always would be ideal, doing it sometimes is a nice thing, and doing it never would be sad but not a big deal.

egelja replied:
`sed -i 's/return 0/return (size_t)0/g' kernel/of/fdt_mem.c`

kjhayes replied:
That changed nothing, there's no `return 0`s in that file

egelja commented on kernel/of/fdt_mem.c:
```diff
@@ -0,0 +1,192 @@
+                fdt_node_name(mem_node),
+                '0' + (unsigned char)i,
+                (void*)address,
+                (void*)(uintptr_t)size
```
```suggestion
                (void*)size
```

we are on C23 and using an intelligent compiler, we don't need to do these old hacks

kjhayes replied:
"We're using an intelligent compiler" is the same line of thought as people misusing `int` and `long`, I'm not certain this cast is needed but it doesn't hurt anything so I'm leaving it.

egelja replied:
It makes readability harder. There's no point to use two casts when one works.

egelja replied:
actually never mind keep it its good

https://stackoverflow.com/a/34291521

kjhayes committed:
Moved device tree files in .gitignore

kjhayes committed:
Added link to Device Tree Spec

kjhayes committed:
Style fixes and redundant elses removed

kjhayes committed:
Added kernel/attribute.h

egelja commented on include/kernel/of/fdt.h:
```diff
@@ -8,7 +8,10 @@
 
-// Bad debugging system (should integrate into Kconfig or delete entirely later)
+// Device Tree Version 17 Specification
+// https://devicetree-specification.readthedocs.io/en/stable/flattened-format.html#figure-device-tree-structure
```
this is SO useful thank you

kjhayes committed:
Fixed bug getting Register Blocks from FDT

kjhayes committed:
Formatting

egelja commented:
You've still got some unresolved comments. 

Sorry for leaving so many 🫠

kjhayes committed:
Handling FDT reserve entry on 32-bit systems

Co-authored-by: Nikola Maruszewski <me@nikola.cx>

kjhayes committed:
Adding check for automatic VLA usage

kjhayes committed:
Merge branch 'fdt_mem' of github.com:riscy-development/riscy-os into fdt_mem

kjhayes committed:
Added KERNEL_ALIGNED attribute

When KERNEL_ALIGNED(b) is applied to an object it changes the alignment of
the struct to "b" bytes. This helps deal with warnings caused by using
KERNEL_PACKED and can help generate new safety warnings when a struct has strange
alignment requirements for some reason in the future.

kjhayes committed:
Removing unused variables and formatting

kjhayes committed:
Simplifying #address-cells and #size-cells calculation

egelja commented on include/kernel/of/fdt.h:
```diff
@@ -21,19 +30,24 @@ struct fdt {
     uint32_t token;
     char unit_name[];
-};
+} KERNEL_PACKED KERNEL_ALIGNED(4);
```
```suggestion
} KERNEL_PACKED KERNEL_ALIGNED(sizeof(uint32_t));
```

kjhayes replied:
`sizeof(uint32_t)` will always and forever be 4 bytes, So I'm not against this change I just don't get the logic behind it. This struct needs to be 4 byte aligned regardless of if we represent the fields inside it with `uint32_t`'s or `unsigned char[4]`'s

kjhayes replied:
Being aligned to `void*` in the first one is because the `memresv` table needs to be 4 byte aligned on 32-bit systems and 8 byte aligned on 64-bit systems (which forces the header to be similarly aligned)

egelja replied:
Reason is to make it explicit what you are aligning to. 

By saying `sizeof(uint32_t)`, you're saying "I want this aligned to the size of a `uint32_t`"

By saying `4`, you could have just arbitrarily picked the number because it was nice or something.

kjhayes replied:
I am making it explicit, I need it aligned to exactly `4` independently of `uint32_t` which just happens to have size `4` in this case

kjhayes replied:
I know `sizeof(uint32_t)` will always be `4`, but so will `ceil(M_PI)`, and all we really care about is a 4 byte alignment, not the `uint32_t` type

egelja replied:
alright sure

egelja commented on kernel/mem/boot/free_list.c:
```diff
@@ -0,0 +1,292 @@
+//
+
+// Linked list of free memory regions
+static struct boot_free_region* boot_free_list = NULL;
```
Can you put this next to the struct definition

egelja committed:
build: better check for VLAs

egelja commented on kernel/of/fdt.c:
```diff
@@ -285,7 +414,13 @@ fdt_string_from_offset(struct fdt* fdt, size_t offset)
+#else
+    // If the first character is the terminator, replace it with a debug name
+    // NOTE: different nodes can appear to have the same name with this scheme
+    return *(char*)node->unit_name != '\0' ? node->unit_name : "{DEBUG: Nameless Node}";
```
is `unit_name` not already a `char*`?

kjhayes replied:
Technically no, it's a `char[]` but I'm pretty much certain dereferencing an array like `*unit_name` is allowed, I just didn't feel like double checking it is. (We could also just do `unit_name[0]`) all of them are equivalent afaik

kjhayes replied:
I can change it if it really bothers you but it doesn't bother me.

egelja replied:
Ya actually I'd do `unit_name[0]` it's clear what you mean then

egelja commented on kernel/of/fdt.c:
```diff
@@ -359,8 +494,240 @@ fdt_find_compatible_node(struct fdt* fdt, struct fdt_node* start, const char* co
+            size_t prop_len = fdt_prop_val_len(prop);
+            void* prop_val = fdt_prop_val(prop);
+            // Consider equal even if '\0' is omitted in the property
+            if (prop_len == len + 1 || prop_len == len) {
```
```suggestion
            if (prop_len == len - 1 || prop_len == len) {
```

Having `\0` omitted would imply `prop_len` is 1 char shorter

kjhayes replied:
`len` is from `strlen` which already omits the `\0`, so the first case is for it the prop doesn't omit the `\0`

egelja replied:
icic ok

kjhayes committed:
Minor syntax changes

kjhayes committed:
YMerge branch 'fdt_mem' of github.com:riscy-development/riscy-os into fdt_mem

kjhayes committed:
Formatting after Merge

egelja reviewed (approved):
lgtm
